Implement a fallback for redirect urls

matrikular opened the pull request:
#### Introduction

The current implementation of the redirect application (component and plugin) expects a direct match to act uppon. So, if one wants to manually add a "rewrite rule" for e.g. http(s)://www.domain.tld/path/some-alias?some=param the rule would have to be exactly that url.

There are a couple of scenarios where query params could be added to any url, like when using a newsletter tool or if you've implemented certain affiliate mechanisms, where, especially in case of a relaunch of a website, you would have to create a whole bunch of redirect rules for each and every newsletter and / or affliiate link. Seldom this amount of different redirects can or should be handled via rewrite rules in a .htacces file.

This PR will add to the default behavior in a way that, if no specific (and published) redirect rule was found, it keeps looking for a "fallback".

Here is an example (behavior without the patch).:
If no (published) redirect rule for http(s)://www.domain.tld/path/some-alias?some=param exists, it will lead to a 404 status code.

After the patch, the plugin would also find http(s)://www.domain.tld/path/some-alias and would append the missing params automatically to the redirect url.
#### Summary of Changes
- Removed unnecessary comment blocks e.g.`$app = JFactory::getApplication(); or $uri = JUri::getInstance();`are common practice and should not need to be explained.
- "Reduced" the database access, refactored the insert / update methods and added separate try / catch blocks to support different error messages for both operations. I've also changed the error code in the catch block from 404 to 500 since in case of an error there, we're not dealing with a 404 anymore.
#### Testing Instructions
- Create a test article and a menu item for that article.
- Activate all SEF options in the global configuration and enable the redirect plugin in the plugin manager. Also activate the option to collect urls if it isn't already.
- If you click that link, everything should be good. Say, you named your article and the corresponding menu item "Test Article", leaving you with the following url: http://www.domain.tld/test-article
- Now, change /test-article to something that does not exist e.g. http://www.domain.tld/shenanigans. This should leave you with a classic 404 error page.
- Head over to the backend and check if that request was captured. If so, edit the entry and set the "Destination URL" to either http://www.domain.tld/test-article or just /test-article and enable the redirect.
- Visit http://www.domain.tld/shenanigans again. You should now be redirected to http://www.domain.tld/test-article.
- For the last part, visit the same url, but this time, add ?this=wontwork to the url so you end up with something similar to: http://www.domain.tld/shenanigans?this=wontwork
- Apply the patch and visit: http://www.domain.tld/shenanigans?this=willwork You should be redirected to your test article.

matrikular committed:
Implement fallback redirect url

matrikular committed:
Satisfy Travis (CodeSniffer)

mbabker commented on plugins/system/redirect/redirect.php:
```diff
@@ -118,125 +116,135 @@ private static function doErrorHandling($error)
-
-		// If no published redirect was found try with the server-relative URL
-		if (!$link || ($link->published != 1))
+		$dbo = JFactory::getDbo();
```
Core had at one point standardized on using `$db` for all database object references.

matrikular replied:
Though I've changed the lines to use $db instead of $dbo, we should then update / change the CodeSniffer rules since they discourage having variables shorter than three characters.

photodude replied:
@matrikular there is a possibility that we might be able to customize the variable length check with an exception for `$db`. Generally, I think we should keep the standard to at least Three characters and make a few limited exceptions for things like $db. That or we should fix all of the `$db` references to `$dbo` so that it complies with the actual written standard and quite trying to have standards that violate our other standards.

mbabker commented on plugins/system/redirect/redirect.php:
```diff
@@ -118,125 +116,135 @@ private static function doErrorHandling($error)
+			->from($dbo->qn('#__redirect_links'))
+			->where(
+				'('
+				. $dbo->qn('old_url') . ' = ' . $dbo->q($url)
```
Likewise, core had standardized on the long `quoteName` and `quote` method names over the shortcuts.

matrikular replied:
I've also changed the code to use the long method names. I still count 452 matches for "->qn(" within a blank Joomla! installation though.

mbabker replied:
`*sigh*` I'm sure whoever went through the work to standardize that feels their time was wasted now.

mbabker commented on plugins/system/redirect/redirect.php:
```diff
@@ -118,125 +116,135 @@ private static function doErrorHandling($error)
-				->where($db->quoteName('old_url') . ' = ' . $db->quote($currRel));
-			$db->setQuery($query, 0, 1);
-			$link = $db->loadObject();
+			JErrorPage::render(new Exception(JText::_('PLG_SYSTEM_REDIRECT_ERROR_UPDATING_DATABASE'), 500, $e));
```
Why should the fact that querying the database table failed cause a 500 code to be sent in place of a 404?  IMO given that at this point the system's established that it is a 404 error that should be used going forward regardless of whether the error handler hits any errors.

matrikular replied:
The result can never become that expected 404 if a database error occurs. Not only from the seo point of view you would want to know exactly whats wrong with that request.

Imagine the maybe far fetched example of the seo department getting in touch with you, stating that there are NNN 404 you should take a look at. Now, if you head over to your redirect rules nothing will look suspicious, causing you and the seo department to have long discussions about who is wrong and who is right. I'd rather be the guy who tells the seo department; no worries, something is wrong with the database / table, I'm already working on it.

mbabker replied:
It's an argument both ways.  Considering this handler is only used when the system throws a 404 error, it seems a little odd that the handler would change to another code when handling an error unless a critical failure was reached while in the error handling closure.  My gut says that since this handler is only recording additional information before displaying the error page that it should stay a 404 error; the user did hit a page that would result in a 404 but the recording of that data failed.

mbabker commented on plugins/system/redirect/redirect.php:
```diff
@@ -118,125 +116,135 @@ private static function doErrorHandling($error)
+				}
+			}
+		}
+		catch (Exception $e)
```
This should only catch the new `JDatabaseExceptionExecuting` exception object, not all Exceptions.  This catch is too eager because it tries to potentially handle errors it should have no awareness of.

matrikular replied:
Sorry, never was my intention to catch any other exceptions. That part should be outside the try block of course.

matrikular committed:
Change all references from $dbo to $db

I would like to point out, that the promoted CodeSniffer Rules discourage having variable names shorter than three characters.

matrikular committed:
Move foreach outside the try catch block

matrikular committed:
Use long method names for nameQuote and quote instead of qn and q

matrikular committed:
Satisfy Travis (Messed up the last changes)

matrikular committed:
Satisfy Travis (Multiple empty lines)

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 53b427782e268e0c406460af4b9f1b5188b4d76f<br /><br />followed test instructions and worked fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10670">issues.joomla.org/joomla-cms/10670</a>.</sub>

chmst commented:
I have tested this item :white_check_mark: successfully on 53b427782e268e0c406460af4b9f1b5188b4d76f<br /><br />I've tested this successfully. Works as described, very useful!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10670">issues.joomla.org/joomla-cms/10670</a>.</sub>

zero-24 commented:
RTC based on testing thanks :+1: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10670">issues.joomla.org/joomla-cms/10670</a>.</sub>

zero-24 commented:
Sorry the `????` was not expected :smile: I have just commented from my phone and i maybe hit the wrong button ;)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10670">issues.joomla.org/joomla-cms/10670</a>.</sub>

wilsonge commented:
Merged - thanks!
